### PR TITLE
wip on optional

### PR DESCRIPTION
**Pull request by mr-feek**
Taking a stab at #119 , documenting where I left off

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -106,38 +108,6 @@ private function ingestFacadeStubs(
         $registration->addStubFile($cache_dir . 'stubs.stubphp');
     }
 
-    /**
-     * @param \Illuminate\Foundation\Application|\Laravel\Lumen\Application $app
-     * @param \Illuminate\View\Factory $view_factory
-     */
-    private function ingestMetaStubs(
```
Comment (mr-feek) [inline src/Plugin.php]: The meta files generated by ide-helper create a stub for `optional`, which prevents us from properly making our own stub. We can just remove the usage of this for now

File: src/Stubs/helpers.stubphp [hunk 2]
```diff
@@ -57,3 +57,15 @@ function abort_if($boolean, $code, $message = '', array $headers = [])
 {
 
 }
+
+/**
+ * Provide access to optional objects.
+ *
+ * @template TValue
+ * @psalm-param TValue $value
+ * @psalm-return \Illuminate\Support\Optional<TValue>
```
Comment (mr-feek) [inline src/Stubs/helpers.stubphp]: This will need to change to account for the callback in the future

**Comment by mr-feek** [comment]
`AtomicMethodCallAnalyzer` is a beast of a class, but looks like the template params aren't being passed properly

File: src/Stubs/helpers.stubphp (hunk 2 above)
Comment (caugner) [inline src/Stubs/helpers.stubphp]: I guess this should work to account for the callback:

```suggestion
 * @template TValue
 * @template TResult
 * @template TCallback of null|callable(TValue): TResult
 * @psalm-param TValue $value
 * @psalm-param TCallback $callback
 * @psalm-return (TCallback is null ? \Illuminate\Support\Optional<TValue> : (TValue is null ? null : TResult))
```

**Pull request merged**